### PR TITLE
[Zen2] Introduce auto_shrink_voting_configuration setting

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today we allow the user to set the minimum size of a voting configuration. On
reflection we would rather this was simply '3' where possible, and we can use
the retirement API to control the removal of nodes more explicitly.

This change replaces the old reconfigurator setting with a new one,
`cluster.auto_shrink_voting_configuration`, which determines whether
Elasticsearch should automatically remove nodes from the voting configuration
or not.

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/cluster/coordination/CoordinatorTests.java [hunk 1]
```diff
@@ -167,8 +165,8 @@ public void testExpandsConfigurationWhenGrowingFromOneToThreeNodesAndShrinksOnFa
         {
             final ClusterNode newLeader = cluster.getAnyLeader();
             final VotingConfiguration lastCommittedConfiguration = newLeader.getLastAppliedClusterState().getLastCommittedConfiguration();
-            assertThat(lastCommittedConfiguration + " should be 1 node", lastCommittedConfiguration.getNodeIds().size(), equalTo(1));
-            assertFalse(lastCommittedConfiguration.getNodeIds().contains(disconnect1.getId()));
+            assertThat(lastCommittedConfiguration + " should be all nodes", lastCommittedConfiguration.getNodeIds(),
+                equalTo(cluster.clusterNodes.stream().map(ClusterNode::getId).collect(Collectors.toSet())));
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/cluster/coordination/CoordinatorTests.java]: test name is not correct anymore? it says "AndShrinksOnFailure"

File: server/src/test/java/org/elasticsearch/cluster/coordination/CoordinatorTests.java [hunk 2]
```diff
@@ -179,10 +177,10 @@ public void testExpandsConfigurationWhenGrowingFromThreeToFiveNodesAndShrinksOnF
 
         final ClusterNode leader = cluster.getAnyLeader();
 
-        logger.info("setting fault tolerance to 1");
-        leader.submitSetMasterNodesFailureTolerance(1);
+        logger.info("setting auto-shrink reconfiguration to true");
+        leader.submitSetAutoShrinkVotingConfiguration(true);
         cluster.stabilise(DEFAULT_CLUSTER_STATE_UPDATE_DELAY);
-        assertThat(CLUSTER_MASTER_NODES_FAILURE_TOLERANCE.get(leader.getLastAppliedClusterState().metaData().settings()), equalTo(1));
+        assertTrue(CLUSTER_AUTO_SHRINK_VOTING_CONFIGURATION.get(leader.getLastAppliedClusterState().metaData().settings()));
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/cluster/coordination/CoordinatorTests.java]: maybe also update the test method name to say that it only shrinks back to 3?

**Review by ywelsch — APPROVED** [review]
LGTM

File: server/src/test/java/org/elasticsearch/cluster/coordination/CoordinatorTests.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/cluster/coordination/CoordinatorTests.java]: The name of the test that contains this line is `testExpandsConfigurationWhenGrowingFromOneNodeToThreeButDoesNotShrink` which seems right to me?

File: server/src/test/java/org/elasticsearch/cluster/coordination/CoordinatorTests.java (hunk 2 above)
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/cluster/coordination/CoordinatorTests.java]: Ok I added `BackToThree`.

File: server/src/test/java/org/elasticsearch/cluster/coordination/CoordinatorTests.java (hunk 1 above)
Comment (ywelsch) [reply server/src/test/java/org/elasticsearch/cluster/coordination/CoordinatorTests.java]: Github UI fooled me

**Pull request merged**